packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +135,186 @@ function createProcessedStreamGenerator(
+   * | 4 bytes            | 4 bytes | 4 bytes        | 4 bytes     | 4 bytes    | 4 bytes     | 4 bytes                    | 4 bytes          | ...   |
+   *
+   * After header, it comes with the real video frame data.
+   * At the end of the texture, it comes with the metadata. The metadata section can contain multiple types of metadata.
+   * Each type of metadata is called a stream. The section is in the following format:
+   * | stream1.id | stream1.dataOffset | stream1.dataSize | stream2.id | stream2.dataOffset | stream2.dataSize | ... | stream1.data | stream2.data | ... |
+   * | :---:      | :---:              | :---:            |  :---:     |  :---:             |  :---:           |:---:|  :---:       | :---:        |:---:|
+   * | 4 bytes    | 4 bytes            | 4 bytes          | 4 bytes    | 4 bytes            | 4 bytes          | ... | ...          | ...          | ... |
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  private extractVideoFrameAndMetadata = async (
+    texture: VideoFrame,
+  ): Promise<{ videoFrame: VideoFrame; metadata: { audioInferenceResult?: Uint8Array } }> => {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    if (texture.format !== 'NV12') {
+      this.notifyError('Unsupported video frame format');
+      throw new Error('Unsupported video frame format');
+    }
+
+    // Identifier for the texture layout, which is the 4-byte ASCII string "oti1" hardcoded by the host
+    // (oti1 stands for "one texture input version 1")
+    const ONE_TEXTURE_LAYOUT_ID = 0x6f746931;
+    // Stream id for audio inference metadata, which is the 4-byte ASCII string "1dia" hardcoded by the host
+    // (1dia stands for "audio inference data version 1")
+    const AUDIO_INFERENCE_RESULT_STREAM_ID = 0x31646961;
+
+    // The rectangle of pixels to copy from the texture. The first two rows are the header.
+    const headerRect = { x: 0, y: 0, width: texture.codedWidth, height: 2 };
+    // allocate buffer for the header
+    // it's in NV12 format (https://learn.microsoft.com/en-us/windows/win32/medfound/recommended-8-bit-yuv-formats-for-video-rendering#nv12), so the
+    // number bytes of the header is (the size of the Y plane + the size of the UV plane)
+    // (headerRect.width * headerRect.height) + (headerRect.width * headerRect.height) / 2
+    const headerBuffer = new ArrayBuffer((headerRect.width * headerRect.height * 3) / 2);
+    await texture.copyTo(headerBuffer, { rect: headerRect });
+    const headerDataView = new Uint32Array(headerBuffer);
+    // headerDataView contains the following data:
+    // 0: oneTextureLayoutId
+    // 1: version
+    // 2: frameRowOffset
+    // 3: frameFormat
+    // 4: frameWidth
+    // 5: frameHeight
+    // 6: multiStreamHeaderRowOffset
+    // 7: multiStreamCount
+    if (headerDataView.length < 8) {
+      this.notifyError('Invalid video frame header');
```

Comment:
everywhere you're throwing errors with duplicated strings, please only define the strings once.